### PR TITLE
Leader election

### PR DESCRIPTION
**Pull request by ymmt2005**
This pull request intends to enable leader election for CSI controller.

Leader election can be implemented with [controller-runtime](https://godoc.org/sigs.k8s.io/controller-runtime/pkg/controller).
To do so, programs should be re-organized as follows:

- `csi-topolvm` is divided and merged into `topolvm-controller` and `topolvm-node`.
- `topolvm-hook` is merged into `topolvm-controller`.
- `topolvm-controller` enables leader election.

The source codes are also re-organized as a single kubebuilder project for simplicity.

### CODE REVIEW

File: driver/k8s/logical_volume.go [hunk 1]
```diff
@@ -165,27 +127,23 @@ func (s *logicalVolumeService) CreateVolume(ctx context.Context, node string, na
 
 func (s *logicalVolumeService) DeleteVolume(ctx context.Context, volumeID string) error {
 	lvList := new(topolvmv1.LogicalVolumeList)
-	err := s.mgr.GetClient().List(ctx, lvList,
-		client.MatchingField(indexFieldVolumeID, volumeID))
+	err := s.List(ctx, lvList, client.MatchingField(indexFieldVolumeID, volumeID))
```
Comment (zoetrope) [inline driver/k8s/logical_volume.go]:
```suggestion
	err := s.List(ctx, lvList, client.MatchingFields{indexFieldVolumeID: volumeID})
```

`client.MatchingField` is deprecated.

File: driver/k8s/logical_volume.go [hunk 2]
```diff
@@ -165,27 +127,23 @@ func (s *logicalVolumeService) CreateVolume(ctx context.Context, node string, na
...
 
-	return s.mgr.GetClient().Delete(ctx, &lvList.Items[0])
+	return s.Delete(ctx, &lvList.Items[0])
 }
 
 func (s *logicalVolumeService) getLogicalVolumeListByVolumeID(ctx context.Context, volumeID string) (*topolvmv1.LogicalVolumeList, error) {
 	lvList := new(topolvmv1.LogicalVolumeList)
-	err := s.mgr.GetClient().List(ctx, lvList,
-		client.MatchingField(indexFieldVolumeID, volumeID))
+	err := s.List(ctx, lvList, client.MatchingField(indexFieldVolumeID, volumeID))
```
Comment (zoetrope) [inline driver/k8s/logical_volume.go]:
```suggestion
	err := s.List(ctx, lvList, client.MatchingFields{indexFieldVolumeID: volumeID})
```

File: config/samples/topolvm_v1_logicalvolume.yaml [hunk 3]
```diff
@@ -4,6 +4,4 @@ metadata:
   name: logicalvolume-sample
 spec:
   # Add fields here
-  size: 10Gi
-  node_name: localhost
-  name: pvc-1234
+  foo: bar
```
Comment (zoetrope) [inline config/samples/topolvm_v1_logicalvolume.yaml]: Is this correct?

Comment (ymmt2005) [reply config/samples/topolvm_v1_logicalvolume.yaml]: It is an automatically generated sample file.
Removing.

File: docs/design.md [hunk 4]
```diff
@@ -15,82 +15,104 @@ Goals
...
+`topolvm-scheduler` is a [scheduler extender](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/scheduling/scheduler_extender.md) to extend the
+standard Kubernetes scheduler for TopoLVM.
+
+`topolvm-controller` implements CSI controller service.  It also works as
+a custom Kubernetes controller to implement dynamic volume provisioning and
+resource cleanups.
+
+`topolvm-node` implements CSI node service.  It also works as a custom
+Kubernetes controller to implement dynamic-volume provisioning at node side.
+All other services are implemented in a program called `csi-topolvm`.
```
Comment (zoetrope) [inline docs/design.md]: Does `csi-topolvm` still exist?

Comment (ymmt2005) [reply docs/design.md]: No.  Will fix.

File: docs/crd-logical-volume.md [hunk 5]
```diff
@@ -1,51 +1,45 @@
...
+| `code`     | uint32 | gRPC error code. If there is no error, `0` is assigned.                            |
+| `message`  | string | Error message.                                                                     |
 
-When `LogicalVolume` resource has been registered, `status.volumeID` is empty.
-Reconciler attempts to create logical volume in LVM.
+Lifecycle
+---------
 
-If Reconciler created logical volume, it set `status.volumeID`.
+Initially, `status.volumeID` is empty.  It is set by `csi-topolvm` on target nodes
```
Comment (zoetrope) [inline docs/crd-logical-volume.md]:
```suggestion
Initially, `status.volumeID` is empty.  It is set by `topolvm-node` on target nodes
```

File: docs/crd-logical-volume.md [hunk 6]
```diff
@@ -1,51 +1,45 @@
...
+after it creates an LVM logical volume.
 
-When `LogicalVolume` is deleted, Finalizer is invoked.
-Finalizer deletes a corresponding logical volume in LVM.
+`LogicalVolume` is created with a [finalizer](https://kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions/#finalizers).
+When a `LogicalVolume` is being deleted, `csi-topolvm` on the target node deletes
```
Comment (zoetrope) [inline docs/crd-logical-volume.md]:
```suggestion
When a `LogicalVolume` is being deleted, `topolvm-node` on the target node deletes
```

**Pull request merged**